### DISCUSSION TITLE
Do i need for racechono, harrys Laptimer etc. Taskscheduler?

### DISCUSSION THREAD

**Discussion by R1Snake** [renatobo/bonogps#28] [Q&A]
Hi,

i am using an esp32 with NEO 8M and without a battery.

The ESP is powered via ignition power of my bike so I don't need deepsleep, battery information etc.

Do I need the Taskscheduler for Racechrono or another racing app? If not I would disable this functionality.

**Comment by renatobo** [comment]
Hi R1Snake!

Please note that a M8N might be limited to 5Hz updates unless you enable only 1 satellite constellation (which is essentially GPS only) then you can get to 10Hz (see the [hardware/GPS](https://github.com/renatobo/bonogps/tree/main/hardware/GPS#q-m8n-m8q-m8u--which-one) section)

**A note on battery support**
Battery support is optional and activated (and compiled in the code) only when the macro `SHOWBATTERY` is defined (right now I only tested on the Lolin D32 Pro).
Not requiring a battery makes for a much easier hardware configuration indeed: be aware of one limitation which is tied to the cold start of the GPS unit at the beginning of your day when it requires some time (even minutes) to find satellites.

**TaskScheduler**

TaskScheduler is optional and it supports

1. status LED blinking: off, slow blink, fast blink to show WiFi status
2. periodic polling GSA and GSV information. These are not actually required for GPS positioning
3. power saving
4. battery status notifications
2. disabling OTA updates after some configured time

Number 2 above is the only one that changes data sent to your app of choice, the other are helper functions that don't impact base GPS functionality.

You can undefine the macro `TASK_SCHEDULER` to remove it.

A couple of questions

1. Is your goal simpler code, smaller binary, or else?
2. what apps are you using and which connection method (BLE,BTSPP,WiFi) ?

**Comment by R1Snake** [comment]
1. My goal ist to learn with your code and create libraries to train the handling with objects.
2. I am using racechrono.

FYI: I updated the code and created some libraries and everything looks good ( I am proud of myself). But I have found a bug (new issue created) and I get the message "No more processes" when I am connected with Wifi, Racechrono is active and I open the browser.

I have a question. I took a look into the DIY Racechrono code and he uses BLE instead of BT SPP. Why do you use SPP? 
I think about Can Bus functionality which is not supported with SPP or?

**Reply by renatobo** [reply]
Thanks for bug report!
I have encountered the **"no more processes"** as well in the serial log and it had to with the webserver component, I noticed it's more evident when you compile with debugging and for pages with more content.
I did some work logged into #2 and there is a branch still active using a different webserver that uses async TCP - the complication became that the tcp/ip client used for GPS data could not work with the async tcp/ip stack so I did not proceed (as some optimization on the HTML static code helped).
If it gets work, there will be work needed. It might be a memory leak, a library bug, inappropriate usage of libraries - after a few hours I put that aside as there was a decent working alternative.

For **Racechrono the choice of BT SPP** depended on

- support of BLE on Android is supposedly not reliable (not my specific experience but it's reported in many places)
- BT SPP has higher bandwidth and its management is simpler

I did not even attempt BLE, do you see any benefit?

**Reply by R1Snake** [reply]
In my code i work with RTOS Tasks because the handling is easier if you work with libraries.

xTaskCreatePinnedToCore(
                this->pollGSAGSVInfo,
                "Pull-GSA-GSV_Information",
                1500,
                (void *)this,
                1,
                &taskHandler,
                0);

My fault was the size of the task. I shrinked it to 1500 and this fixed the behavior. I think the Task you use needs to much space. Can you shrink the space of your Tasks? If yes please try this

**Reply by R1Snake** [reply]
Next thing i found out is that you are using Strings. I think this is not the best way, what do you think?

In the Handle menu method I log at the end the Free heap and I shrinks more call after call. I will write the code new to use char arrays and hope this fix the behavior

**Reply by R1Snake** [reply]
The reason i asked for BLE for racechrono is, that the DIY Code from racechrono works with BLE and describes how the data has to look if you want to send Can-BUS Data. 

I think the Racechrono app has a NMEA Decoder integrated but for Can-Bus this feature is not provided

**Reply by renatobo** [reply]
> Next thing i found out is that you are using Strings. I think this is not the best way, what do you think?
> 
> In the Handle menu method I log at the end the Free heap and I shrinks more call after call. I will write the code new to use char arrays and hope this fix the behavior

Yes Strings are used when its convenience was greater than the savings. Char arrays might be more efficient in some cases: most of the largest text parts are save in PROGMEM in any case and served directly from there. There is some a limited number of options for types to be used with the simple web server library used.

**Reply by renatobo** [reply]
> The reason i asked for BLE for racechrono is, that the DIY Code from racechrono works with BLE and describes how the data has to look if you want to send Can-BUS Data.
> 
> I think the Racechrono app has a NMEA Decoder integrated but for Can-Bus this feature is not provided

I looked again at the RaceChrono example (the one linked in the main README.md file https://github.com/aollin/racechrono-ble-diy-device )
In the meanwhile, let me admit I mostly use HLT so I have not been investing a great deal of time in the other options: I was happy they work, my second preferred option is TrackAddict.

A few considerations

- all current connection methods expect NMEA parsing/decoding on the receiving end, while the example in https://github.com/aollin/racechrono-ble-diy-device requires decoding on the esp32. Not a huge problem (it's actually in there, disabled with macro `NEED_NEOGPS`). Doing this is "high maintenance" as it becomes very dependent on your specific configuration, so I opted for something generic that works across the board.
- CanBus is not a standard format: in my case I opted to use Harry Lap Timer scripting to import data from the ECU with a proprietary binary protocol (see below)
- author describes the BLE example as "no way ready as a product" status: this is a great put off if he does not stand behind it :))
- Using several characteristics in BLE (2 for GPS and 2 for CanBus) is clean and optimal as a design, but this is where the nature of BLE hits its practical use as its inefficiency limits its bandwidth compared to BT-SPP

Considering these points and working under a constraint of the amount of free time, I don't see too many benefits right now in refactoring the BTSPP option into BTE, unless you are trying to do exactly as the example including CanBus: the end result in collecting GPS data is going to be the same so I don't see the added value.

Obviously I am not factoring in the willingness to make it work :) I'll help as I can if you decide to try, yet please consider my priority is on the existing code or code that adds functionalities or solve issues.

For CanBus, you might want to take a look at the CanBus data collector for HLT I have been working on which is at https://github.com/renatobo/DucatiPanigaleCanBus which collects CanBus data.
If that becomes the only option to use CanBus data in RaceChrono, perhaps it might become useful: I am stuck on HTL at the moment so I doubt I will take this on myself.

Thanks for the good pointers and suggestions!
